formatted WSI.
[2025-01-16 16:43:20.238247] [0x00007f7fd40eea00] [debug]    
[2025-01-16 16:43:20.238253] [0x00007f7fd40eea00] [debug]   Level Count: 1
[2025-01-16 16:43:20.238260] [0x00007f7fd40eea00] [debug]   Level: 0
[2025-01-16 16:43:20.238271] [0x00007f7fd40eea00] [debug]   Dimensions Level[0]: 2220, 2967
[2025-01-16 16:43:20.238276] [0x00007f7fd40eea00] [debug]   Level[0] frames:89, 119
[2025-01-16 16:43:20.238280] [0x00007f7fd40eea00] [debug]   Level: 1
[2025-01-16 16:43:20.238289] [0x00007f7fd40eea00] [debug]   Dimensions Level[1]: 69, 92
[2025-01-16 16:43:20.238328] [0x00007f7fd40eea00] [debug]   Level[1] frames:3, 4
[       OK ] getSlideLevelDim.progressive_virtual_levels (11 ms)
[ RUN      ] getSlideLevelDim.progressive_virtual_level_start
[2025-01-16 16:43:20.238719] [0x00007f7fd40eea00] [info]    dicomization is started
[2025-01-16 16:43:20.251003] [0x00007f7fd40eea00] [info]    Reading aperio formatted WSI.
[2025-01-16 16:43:20.251353] [0x00007f7fd40eea00] [debug]    
[2025-01-16 16:43:20.251505] [0x00007f7fd40eea00] [debug]   Level Count: 1
[2025-01-16 16:43:20.251654] [0x00007f7fd40eea00] [debug]   Level: 0
[2025-01-16 16:43:20.251822] [0x00007f7fd40eea00] [debug]   Dimensions Level[0]: 69, 92
[2025-01-16 16:43:20.251980] [0x00007f7fd40eea00] [debug]   Level[0] frames:3, 4
[       OK ] getSlideLevelDim.progressive_virtual_level_start (13 ms)
[----------] 4 tests from getSlideLevelDim (48 ms total)

[----------] 3 tests from zlibWrapper
[ RUN      ] zlibWrapper.small_mem
[       OK ] zlibWrapper.small_mem (0 ms)
[ RUN      ] zlibWrapper.large_mem
[       OK ] zlibWrapper.large_mem (940 ms)
[ RUN      ] zlibWrapper.empty
[       OK ] zlibWrapper.empty (0 ms)
[----------] 3 tests from zlibWrapper (941 ms total)

[----------] Global test environment tear-down
[==========] 79 tests from 24 test suites ran. (3492 ms total)
[  PASSED  ] 79 tests.
+ set +e
./endToEnd/endToEndTest.sh: line 39: pip: command not found
Test 1
[2025-01-16 16:43:21.230665] [0x00007fa7ca9afa00] [info]    dicomization is started
[2025-01-16 16:43:21.230755] [0x00007fa7ca9afa00] [warning] studyId is going to be generated
[2025-01-16 16:43:21.230761] [0x00007fa7ca9afa00] [warning] seriesId is going to be generated
[2025-01-16 16:43:21.230766] [0x00007fa7ca9afa00] [warning] threads parameter is less than 1, consuming all avalible threads
[2025-01-16 16:43:21.230769] [0x00007fa7ca9afa00] [warning] batch parameter is not set, batch is unlimited
[2025-01-16 16:43:21.242121] [0x00007fa7ca9afa00] [info]    Reading aperio formatted WSI.
[2025-01-16 16:43:21.586771] [0x00007fa7c802c6c0] [warning] JPEG 2000: Image size is smaller than 2^(numresolution - 1); Changing numresolution from: 6 to: 4 to meet encoder requirments.
[2025-01-16 16:43:21.689596] [0x00007fa7ca9afa00] [info]    dicomization is done
Traceback (most recent call last):
  File "/workspace/./endToEnd/strip_pixel_data.py", line 3, in <module>
    import pydicom
ModuleNotFoundError: No module named 'pydicom'
ERROR
ERROR: build step 0 "debian:12.4-slim" failed: step exited with non-zero status: 1

--- a/Dockerfile
+++ b/Dockerfile
@@ -1,7 +1,22 @@
 FROM debian:12.9-slim
 
-COPY . /workspace
+WORKDIR /workspace/
 
-WORKDIR /workspace
+COPY cloud_build/ /workspace/cloud_build/
 
-RUN bash cloud_build/debianBuild.sh
+COPY CMakeLists.txt /workspace/
+
+COPY src/ /workspace/src/
+
+COPY tests/ /workspace/tests/
+
+# This keeps a copy of build/wsi2dcm & libwsi2dcm.so as it is used by endToEndTest.sh
+RUN bash /workspace/cloud_build/debianBuild.sh
+
+RUN \
+    cp build/wsi2dcm build/libwsi2dcm.so . && \
+    rm -r build && \
+    mkdir build && \
+    cp ./wsi2dcm libwsi2dcm.so build/
+
+CMD ["wsi2dcm"]

--- a/Dockerfile
+++ b/Dockerfile
@@ -1,0 +1,7 @@
+FROM debian:12.9-slim
+
+COPY . /workspace
+
+WORKDIR /workspace
+
+RUN bash cloud_build/debianBuild.sh

--- a/README.md
+++ b/README.md
@@ -85,11 +85,16 @@ Drop first row and column of the source image in order to workaround bug https:/
 
 ## Compiling from source
 
-If you're using Ubuntu, run the following command to download the dependencies and build the tool:
-
+Using docker is the most reliable way to build from source.
 
 ```shell
-sudo ./cloud_build/ubuntuBuild.sh
+docker build .
+```
+
+If you're using Debian 12, run the following command to download the dependencies and build the tool:
+
+```shell
+sudo ./cloud_build/debianBuild.sh
 ```
 
 Otherwise, follow these steps:

--- a/cloud_build/debianBuild.sh
+++ b/cloud_build/debianBuild.sh
@@ -24,14 +24,18 @@
 # 9: install jsoncpp
 # 10: build
 
+set -ex
+
 #1
 apt-get update 
 apt-get upgrade -y
-apt-get install -y \
+apt-get install -y --no-install-recommends \
+    ca-certificates \
     apt-utils \
     wget \
     build-essential \
     cmake \
+    g++ \
     unzip \
     libgtest-dev \
     libxml2-dev \

--- a/cloud_build/debianBuild.sh
+++ b/cloud_build/debianBuild.sh
@@ -33,8 +33,11 @@ apt-get install -y --no-install-recommends \
     ca-certificates \
     apt-utils \
     wget \
-    build-essential \
+    dpkg-dev \
     cmake \
+    make \
+    meson \
+    ninja-build \
     g++ \
     unzip \
     libgtest-dev \
@@ -50,7 +53,7 @@ apt-get install -y --no-install-recommends \
     libglib2.0-dev      
 #2
 # installing in /workspace
-apt-get install -y nasm
+apt-get install -y --no-install-recommends nasm
 wget -O libjpeg_turbo.zip https://github.com/libjpeg-turbo/libjpeg-turbo/archive/refs/tags/3.1.0.zip > /dev/null
 unzip libjpeg_turbo.zip > /dev/null
 rm libjpeg_turbo.zip
@@ -63,7 +66,7 @@ cd ..
 cd ..
 rm -rf libjpeg-turbo-3.1.0
 #3
-apt-get install -y liblcms2-dev libzstd-dev libwebp-dev
+apt-get install -y --no-install-recommends liblcms2-dev libzstd-dev libwebp-dev
 wget -O v2.5.3.zip  https://github.com/uclouvain/openjpeg/archive/refs/tags/v2.5.3.zip > /dev/null
 unzip v2.5.3.zip > /dev/null
 mkdir -p ./openjpeg-2.5.3/build
@@ -123,9 +126,6 @@ cd boost_1_87_0
 cd ..
 rm -rf boost_1_87_0
 # 8
-apt-get update
-apt-get install -y python3-full pip libjpeg-dev ninja-build
-pip3 install --break-system-packages meson
 wget -O openslide-4.0.0.tar.gz https://github.com/openslide/openslide/archive/refs/tags/v4.0.0.tar.gz > /dev/null
 tar xvzf openslide-4.0.0.tar.gz > /dev/null
 rm openslide-4.0.0.tar.gz

--- a/endToEnd/endToEndTest.sh
+++ b/endToEnd/endToEndTest.sh
@@ -29,10 +29,16 @@ compare () {
 }
 
 
-if [ "$1" !=  "SKIP_BUILD_ENVIRONMENT" ]
+if [ "$1" != "SKIP_BUILD_ENVIRONMENT" ]
 then
   #building binaries
   bash ./cloud_build/debianBuild.sh
+fi
+
+if [ ! -x pip ]; then
+  apt-get install -y --no-install-recommends \
+    python3 \
+    python3-pip
 fi
 
 export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/usr/local/lib:/usr/local/lib/x86_64-linux-gnu:/usr/local/lib/aarch64-linux-gnu

